sqlsimplekv/keyvalue.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package sqlsimplekv
+
+import (
+	"context"
+	"database/sql"
+	sqldriver "database/sql/driver"
+	"fmt"
+	"time"
+
+	errgo "gopkg.in/errgo.v1"
+
+	"github.com/juju/simplekv"
+)
+
+// NewStores returns a new Stores instance that uses the
+// given sql database for storage, generating SQL with the
+// given driver (currently only "postgres" is supported).
+//
+// The data will be stored in a table with the given name
+// (other SQL artificacts may also be created using the name as a prefix).
+func NewStore(driverName string, db *sql.DB, tableName string) (simplekv.Store, error) {
+	if driverName != "postgres" {
+		return nil, errgo.Newf("unsupported database driver %q", driverName)
+	}
+	driver, err := newPostgresDriver(db, tableName)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot initialise database")
+	}
+	return &kvStore{
+		tableName: tableName,
+		db:        db,
+		driver:    driver,
+	}, nil
+}
+
+// A kvStore implements simplekv.Store.
+type kvStore struct {
+	db        *sql.DB
+	driver    *driver
+	tableName string
+}
+
+// Context implements simplekv.Store.Context.
+func (s *kvStore) Context(ctx context.Context) (context.Context, func()) {
+	return ctx, func() {}
+}
+
+type keyValueParams struct {
+	argBuilder
+
+	TableName string
+	Store     string
+	Key       string
+	Value     []byte
+	Expire    nullTime
+	Update    bool
+}
+
+// Get implements simplekv.Store.Get by selecting the blob with the
+// given key from the table.
+func (s *kvStore) Get(ctx context.Context, key string) ([]byte, error) {
+	v, err := s.get(ctx, s.db, key, false)
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Is(simplekv.ErrNotFound))
+	}
+	return v, nil
+}
+
+// get is like Get except that it operates on a general queryer value.
+// If forUpdate is true, it takes out a lock on the given key so that a subsequent
```

Comment:
row level locking is only available on certain databases. A more generic structure might need a more complex logic. Although let's not solve problems we don't yet have.

Reply:
Well, it takes a lock out on *at least* that key, even when the db doesn't support row-level locking, so technically I think it's still accurate.